src/Orleans.Runtime/MembershipService/MembershipTableFactory.cs

```diff
@@ -7,38 +7,18 @@ namespace Orleans.Runtime.MembershipService
 {
     internal class MembershipTableFactory
     {
-        private readonly IServiceProvider serviceProvider;
-        private readonly AsyncLock initializationLock = new AsyncLock();
         private readonly ILogger logger;
         private IMembershipTable membershipTable;
 
-        public MembershipTableFactory(IServiceProvider serviceProvider, ILogger<MembershipTableFactory> logger)
+        public MembershipTableFactory(IMembershipTable membershipTable, ILogger<MembershipTableFactory> logger)
         {
-            this.serviceProvider = serviceProvider;
+            this.membershipTable = membershipTable;
             this.logger = logger;
         }
 
-        internal async Task<IMembershipTable> GetMembershipTable()
+        internal IMembershipTable GetMembershipTable()
         {
-            if (membershipTable != null) return membershipTable;
```

Comment:
please take a look of this code. To me the lock and all those null handling seems unnecessary since membership is not injected from DI and should only be initialized in `MembershipOracle.Start` method. 

ManagementGrain calls `GetMembershipTable` for hosts information. But it shouldn't call it before it is initialized nor initialize it before `MembershipOracle.Start`. So I think this change is safe. But do take a look in case I missed anything.

Reply:
It seems that the previous pattern always relied on the factory to create and initialize the membership table, whereas now the factory can create and return an uninitialized table.  While this may work in the current call patterns it seems fragile.  Am I misunderstanding what I'm seeing?

Reply:
fragile as in caller for GetMembershipTable may get one instance which hasn't been initialized yet? yes it is true. 
Or am I missing other fragile points?

Reply:
But I think the previous pattern is fragile too. So any caller calls GetMembershipTable will initialize the table. But to initialize the table,  it need certain dependency to be initialized too. Previous pattern doesn't handle this well. 

New pattern doesn't initialize the table in GetMembershipTable() call, it holds the risk of returning a uninitialized table. If this uninitialized table get called , then the table implementation should throw UnInitializedException or some sort. I think this should be the correct behavior. Do you disagree?

Reply:
There are many things wrong with the current pattern.
Factory
* A factory should create instances in a usable form, initializing it if it's necessary
* In general factories should not cache instances of what they create.  Caches or singletons should do that.

Basically this is backwards, the factory does not create the membership table, it gets it from the container and initializes it, so it's not a factory at all..?

I don't think we need or should use a factory in this case.  I think membership tables should register in the container as an IMembershipTable and as a lifecycle participant and initialize themselves in the startup process.  If they are called prior to initialization, they should throw an error.

If anything calls the membership service prior to it's initialization, it would be due to misconfiguration, and users could configure the membership to start at an earlyer stage, or start dependent systems at later stages.

Open to other patterns, but this PR seems to just exchange a questionable approach for a different questionable approach, which admitadly is far less work than what I'd suggested above.

Reply:
offline discussion agreed to remove the factory since it is useless at this point.

Reply:
sounds good to me, the factory was a crutch. Just make sure that everything still works when not using *any* `IMembershipTable` implementation and replacing the `IMembershipOracle` implementation instead (Service Fabric case)